src/character.cpp

```diff
@@ -13043,6 +13043,109 @@ void Character::search_surroundings()
     }
 }
 
+bool Character::wield_new( item it )
+{
+    return wield( item_location( *this, &it ), /*remove_old=*/false );
+}
+
+bool Character::wield( item_location loc, bool remove_old )
+{
+    if( !loc ) {
+        add_msg_if_player( _( "No item." ) );
+        return false;
+    }
+
+    // for [w] -> unwield
+    if( has_wield_conflicts( *loc ) ) {
+        const bool is_unwielding = is_wielding( *loc );
+        const auto ret = can_unwield( *loc );
+
+        if( !ret.success() ) {
+            add_msg_if_player( m_info, "%s", ret.c_str() );
+            return false;
+        }
+
+        if( !unwield() ) {
+            return false;
+        }
+
+        if( is_unwielding ) {
+            if( !martial_arts_data->selected_is_none() ) {
+                martial_arts_data->martialart_use_message( *this );
+            }
+            return true;
+        }
+    }
+
+    item_location wielded = get_wielded_item();
+    if( wielded && wielded->has_item( *loc ) ) {
+        add_msg_if_player( m_info,
+                           _( "You need to put the bag away before trying to wield something from it." ) );
+        return false;
+    }
+
+    if( !can_wield( *loc ).success() ) {
+        return false;
+    }
+
+    bool combine_stacks = wielded && loc->can_combine( *wielded );
+    if( !combine_stacks && !unwield() ) {
+        return false;
+    }
+
+    cached_info.erase( "weapon_value" );
+
+    if( is_avatar() && !avatar_action::check_stealing( *this, *loc ) ) {
+        return false;
+    }
+
+    // Wielding from inventory is relatively slow and does not improve with increasing weapon skill.
+    // Worn items (including guns with shoulder straps) are faster but still slower
+    // than a skilled player with a holster.
+    // There is an additional penalty when wielding items from the inventory whilst currently grabbed.
+
+    bool worn = is_worn( *loc );
+    const int mv = loc.obtain_cost( *this );
+
+    if( worn ) {
+        loc->on_takeoff( *this );
+    }
+
+    add_msg_debug( debugmode::DF_AVATAR, "wielding took %d moves", mv );
+    mod_moves( -mv );
+
+    if( combine_stacks ) {
+        wielded->combine( *loc );
+    } else {
+        set_wielded_item( *loc );
+    }
+
+    if( remove_old && loc ) {
+        loc.remove_item();
+    }
+
+
+    // set_wielded_item invalidates the weapon item_location, so get it again
+    wielded = get_wielded_item();
+    last_item = wielded->typeId();
+    recoil = MAX_RECOIL;
+
+    wielded->on_wield( *this );
+
+    cata::event e = cata::event::make<event_type::character_wields_item>( getID(), last_item );
+    get_event_bus().send_with_talker( this, &wielded, e );
+
+    if( is_npc() ) {
+        add_msg_if_player_sees( *this, m_info, _( "<npcname> wields a %s." ),  wielded->tname() );
+        as_npc()-> invalidate_range_cache();
+    }
+
+    inv->update_invlet( *wielded );
+    inv->update_cache_with_item( *wielded );
```

Comment:
This kind of stuff should move to npc/avatar overloads in my opinion. It keeps the method shorter and makes it clear which code is relevant for which class. (The last two lines only assuming invlets do nothing for npcs.)